src/crypto.rs

```diff
@@ -381,20 +369,152 @@ impl PrivateKey {
+                "Key {:?} can't be used with scheme {:?}",
+                self.private, s
+            )))?,
+        }
+
+        let value = SignatureValue(self.private.sign(msg).as_ref().into());
+        Ok(Signature {
+            key_id: self.public.key_id().clone(),
+            value,
+        })
+    }
+
+    fn public(&self) -> &PublicKey {
+        &self.public
+    }
+}
+
+/// A structure containing information about an Rsa private key.
+pub struct RsaPrivateKey {
+    private: Arc<RsaKeyPair>,
+    public: PublicKey,
+}
+
+impl RsaPrivateKey {
+    /// Generate RSA key bytes in pkcs8 format.
+    ///
+    /// Note: `openssl` needs to the on the `$PATH`.
+    pub fn keypair() -> Result<Vec<u8>> {
+        let gen = Command::new("openssl")
+            .args(&[
+                "genpkey",
+                "-algorithm",
+                "RSA",
+                "-pkeyopt",
+                "rsa_keygen_bits:4096",
+                "-pkeyopt",
+                "rsa_keygen_pubexp:65537",
+                "-outform",
+                "der",
+            ])
+            .output()?;
+
+        let mut pk8 = Command::new("openssl")
+            .args(&[
+                "pkcs8", "-inform", "der", "-topk8", "-nocrypt", "-outform", "der",
+            ])
+            .stdin(Stdio::piped())
+            .stdout(Stdio::piped())
+            .spawn()?;
+
+        match pk8.stdin {
+            Some(ref mut stdin) => stdin.write_all(&gen.stdout)?,
+            None => return Err(Error::Opaque("openssl has no stdin".into())),
+        };
+
+        Ok(pk8.wait_with_output()?.stdout)
+    }
+
+    fn from_pkcs8(der_key: &[u8], scheme: SignatureScheme) -> Result<Self> {
+        if let SignatureScheme::Ed25519 = scheme {
+            Err(Error::IllegalArgument(
+                "RSA keys do not support the Ed25519 signing scheme".into(),
+            ))?
+        }
+
+        let key = RsaKeyPair::from_pkcs8(der_key)
+            .map_err(|_| Error::Encoding("Could not parse key as PKCS#8v2".into()))?;
+
+        if key.public_modulus_len() < 256 {
+            return Err(Error::IllegalArgument(format!(
+                "RSA public modulus must be 2048 or greater. Found {}",
+                key.public_modulus_len() * 8
+            )));
+        }
+
+        let pub_key = extract_rsa_pub_from_pkcs8(der_key)?;
+
+        let public = PublicKey::new(
+            KeyType::Rsa,
+            scheme,
+            python_tuf_compatibility_keyid_hash_algorithms(),
+            pub_key,
         )?;
-        let private = PrivateKeyType::Ed25519(key);
+        let private = Arc::new(key);
 
-        Ok(PrivateKey { private, public })
+        Ok(RsaPrivateKey { private, public })
     }
+}
+
+impl PrivateKey for RsaPrivateKey {
+    fn sign(&self, msg: &[u8]) -> Result<Signature> {
+        let rng = SystemRandom::new();
+        let mut buf = vec![0; self.private.public_modulus_len()];
+        let scheme = match &self.public.scheme {
```

Comment:
Same here, we shouldn't need to check the scheme here.

Reply:
We need the match anyway for converting the RSA scheme type. Made the misc. branch an unreachable!()